rgw: add a method to purge all associate keys when removing a subuser

Sandy4999 opened the pull request:
Fixes: #12890

When removing a subuser, make sure all of its keys, including the swift key and possible s3 keys, are also deleted.

Signed-off-by: Sangdi Xu <xu.sangdi@h3c.com>

Sandy4999 commented:
@yehudasa  Need your review please.
It functions okay but there might be two issues if we do this:
1. the op_state (key_type and access_key) is modified during the procedure. We may record the original values and recover them after deleting the keys, but currently I haven't seen the necessity.
2. the parameter --purge-keys would be redundant since we do purge keys whenever removing a user or a subuser.

yehudasa commented:
@Sandy4999 I don't like the whole state changing scheme, can we change the underlying methods or add a method down at RGWUser that will just do it for us without the need to set and re-set the state?

Sandy4999 commented on src/rgw/rgw_user.cc:
```diff
@@ -1029,6 +1029,37 @@ int RGWAccessKeyPool::modify_key(RGWUserAdminOpState& op_state, std::string *err
+  for (; user_kiter != user_info.access_keys.end(); ++user_kiter) {
+    if (user_kiter->second.subuser == subuser_str) {
+      kiter = keys_map->find(user_kiter->first);
+      rgw_remove_key_index(store, kiter->second);
```
Probably better to add 
```
if (kiter != keys_map->end()) {}
```
here, I will do it tomorrow when I get to work.

Sandy4999 commented on src/rgw/rgw_user.cc:
```diff
@@ -1029,6 +1029,37 @@ int RGWAccessKeyPool::modify_key(RGWUserAdminOpState& op_state, std::string *err
+  map<std::string, RGWAccessKey>::iterator user_kiter = user_info.access_keys.begin();
+  for (; user_kiter != user_info.access_keys.end(); ++user_kiter) {
+    if (user_kiter->second.subuser == subuser_str) {
+      kiter = keys_map->find(user_kiter->first);
```
Probably better to add 
```
if (kiter != keys_map->end()) {}
```
here, I will do it tomorrow when I get to work.

Sandy4999 commented:
@yehudasa I add a method for RGWAccessKeyPool to do the purging keys. Please take a look at this, thank you

Sandy4999 commented:
@yehudasa Can you please review the changes? Thanks.

Sandy4999 commented:
@yehudasa I have added the check to make sure user info is populated and subuser is specified, and also the defer parameter. Moreover, I corrected the way to remove the swift key as it requires 'user:subuser' format index. Please review, thank you!

Sandy4999 commented:
@yehudasa Could you provide a bit more detailed explanation about this parameter? Thank you!

yehudasa commented:
@Sandy4999 this parameter is used when we want to stage changes, but not apply them immediately

Sandy4999 commented on src/rgw/rgw_user.cc:
```diff
@@ -1333,12 +1390,10 @@ int RGWSubUserPool::execute_remove(RGWUserAdminOpState& op_state,
-    user->keys.remove(op_state, &subprocess_msg, true);
-  }
+  // always purge all associate keys
+  user->keys.remove_subuser_keys(op_state, &subprocess_msg, defer_user_update);
```
@yehudasa Okay, I have modified the third parameter to be defer_user_update, not always true. Please take a look at it, thank you!
